compiler/rustc_data_structures/src/statistics.rs

```diff
@@ -0,0 +1,164 @@
+//! # Support for collecting simple statistics
+//!
+//! Statistics are useful for collecting metrics from optimization passes, like
+//! the number of simplifications performed. To avoid introducing overhead, the
+//! collection of statistics is enabled only when rustc is compiled with
+//! debug-assertions.
+//!
+//! Statistics are static variables defined in the module they are used, and
+//! lazy registered in the global collector on the first use. Once registered,
+//! the collector will obtain their values at the end of compilation process
+//! when requested with -Zmir-opt-stats option.
+
+use parking_lot::{const_mutex, Mutex};
+use std::fmt::Write as _;
+use std::io::{self, Write as _};
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+
+static COLLECTOR: Collector = Collector::new();
+
+/// Enables the collection of statistics.
+/// To be effective it has to be called before the first use of a statistics.
+pub fn try_enable() -> Result<(), ()> {
+    COLLECTOR.try_enable()
+}
+
+/// Prints all statistics collected so far to the stderr.
+pub fn print() {
+    COLLECTOR.print();
+}
+
+pub struct Statistic {
+    category: &'static str,
+    name: &'static str,
+    initialized: AtomicBool,
+    value: AtomicUsize,
+}
+
+struct Collector(Mutex<State>);
+
+struct State {
+    enabled: bool,
+    stats: Vec<&'static Statistic>,
+}
+
+#[derive(Eq, PartialEq, Ord, PartialOrd)]
+struct Snapshot {
+    category: &'static str,
+    name: &'static str,
+    value: usize,
+}
+
+impl Statistic {
+    pub const fn new(category: &'static str, name: &'static str) -> Self {
+        Statistic {
+            category,
+            name,
+            initialized: AtomicBool::new(false),
+            value: AtomicUsize::new(0),
+        }
+    }
+
+    pub fn name(&self) -> &'static str {
+        self.name
+    }
+
+    pub fn category(&self) -> &'static str {
+        self.category.rsplit("::").next().unwrap()
+    }
+
+    #[inline]
+    pub fn register(&'static self) {
+        if cfg!(debug_assertions) {
+            if !self.initialized.load(Ordering::Acquire) {
+                COLLECTOR.register(self);
+            }
+        }
+    }
+
+    #[inline]
+    pub fn increment(&'static self, value: usize) {
+        if cfg!(debug_assertions) {
+            self.value.fetch_add(value, Ordering::Relaxed);
```

Comment:
Please use `SeqCst` unless it has a performance impact **and** you can prove that it is correct.

Reply:
That would give a wrong impression that some form of synchronization is required here. Atomicity of RMW operations is sufficient for correctness.